guide/src/class/call.md

```diff
@@ -65,3 +65,51 @@ def Counter(wraps):
         return wraps(*args, **kwargs)
     return call
 ```
+
+### What are the atomics for?
+
+A [previous implementation] used a normal `u64`, which meant it required a `&mut self` receiver to update the count:
+
+```rust,ignore
+#[args(args = "*", kwargs = "**")]
+fn __call__(&mut self, py: Python<'_>, args: &PyTuple, kwargs: Option<&PyDict>) -> PyResult<Py<PyAny>> {
+    self.count += 1;
+    let name = self.wraps.getattr(py, "__name__")?;
+
+    println!("{} has been called {} time(s).", name, self.count);
+
+    // After doing something, we finally forward the call to the wrapped function
+    let ret = self.wraps.call(py, args, kwargs)?;
+
+    // We could do something with the return value of
+    // the function before returning it
+    Ok(ret)
+}
+```
+
+The problem with this is that the `&mut self` receiver means PyO3 has to borrow it exclusively,
+ and hold this borrow across the`self.wraps.call(py, args, kwargs)` call. This call returns control to the user's Python code
+ which is free to call arbitrary things, *including* the decorated function. If that happens PyO3 is unable to create a second unique borrow and will be forced to raise an exception.
+
+As a result, something innocent like this will raise an exception:
+
+```py
+@Counter
+def say_hello():
+    if say_hello.count < 2:
+        print(f"hello from decorator")
+        
+say_hello()
+# RuntimeError: Already borrowed
+```
+
+The implementation in this chapter fixes that by never borrowing exclusively; all the methods take `&self` as receivers, of which multiple may exist simultaneously. This requires a shared and threadsafe counter and the easiest way to do that is to use atomics, so that's what is used here.
+
+This shows the dangers of running arbitrary Python code - note that "running arbitrary Python code" can be far more subtle than the example above:
+- Python's asynchronous executor may park the current thread in the middle of Python code, even in Python code that *you* control, and let other Python code run.
+- Dropping arbitrary Python objects may invoke destructors defined in Python (`__del__` methods).
+- Calling Python's C-api (most PyO3 apis call C-api functions internally) may raise exceptions, which may allow Python code in signal handlers to run.
+
+This especially important if you are writing unsafe code; Python code must never be able to cause undefined behavior. You must ensure that your Rust code is in a consistent state before doing any of the above things.
```

Comment:
> Python code must never be able to cause undefined behavior.

I am not sure that this a tenable position as Python code can already cause undefined behaviour (like data races) even without Rust code being involved, or to more succinct: Python code is unsafe/trusted, not safe/verified like safe Rust.

Reply:
I've always thought of the Python <-> C boundary as the same as safe Rust <-> unsafe Rust; C code can never trust python code, and Python code must never be able to cause undefined behavior.

Obviously there are some exceptions like `ctypes`, which is really just an escape hatch on the python side.

Reply:
At least NumPy does not seem handle it that way, e.g.

```python
from threading import Thread

import numpy as np

def increment_array():
    global array

    for i in range(10_000):
        array += 1

array = np.zeros((100, 100))

threads = [Thread(target=increment_array) for _ in range(10)]
for thread in threads: thread.start()
for thread in threads: thread.join()

assert np.sum(array) == 10 * 10_000 * 100 * 100
```

sometimes passes, sometimes fails presumably because NumPy releases the GIL while doing `array += 1`.

I think there are similar issues in Python's standard library where the implementation of types releases the GIL before doing system calls which should also allow constructing such data races, but it is admittedly harder to do. (For example, a simple counter does not suffice as it only creates a logical race if the update is not done "GIL atomically" using e.g. `counter += 1`, but not a data race AFAIU.)

Reply:
> IIRC there's a cap on the number of bytecode operations before the GIL is released to give other threads a chance to make progress.

I don't think this applies here, i.e. the disassembly of `increment_array` is

```
  8           0 LOAD_GLOBAL              0 (range)
              2 LOAD_CONST               1 (10000)
              4 CALL_FUNCTION            1
              6 GET_ITER
        >>    8 FOR_ITER                 6 (to 22)
             10 STORE_FAST               0 (i)

  9          12 LOAD_GLOBAL              1 (array)
             14 LOAD_CONST               2 (1)
             16 INPLACE_ADD
             18 STORE_GLOBAL             1 (array)
             20 JUMP_ABSOLUTE            4 (to 8)

  8     >>   22 LOAD_CONST               0 (None)
             24 RETURN_VALUE
```

meaning that `INPLACE_ADD` which will call into NumPy eventually is a single bytecode operation.

Reply:
Ah I see what you mean; yes, that's scary, looks like `numpy` does indeed allow a data race while doing the increment!

Reply:
NumPy does it when only primitive types are involved which can be access without involving the interpreter (which is the case in the above example), c.f. the [`NPY_METH_REQUIRES_PYAPI` flag](https://github.com/numpy/numpy/blob/26d690e5abe9dc7eb74deb807ea2c4842aa91c70/numpy/core/src/multiarray/array_method.h#L16) and its usage in the [low-level implementation details](https://github.com/numpy/numpy/blob/26d690e5abe9dc7eb74deb807ea2c4842aa91c70/numpy/core/src/multiarray/lowlevel_strided_loops.c.src#L1571).

Reply:
@adamreichold do you strongly disagree with the wording? I personally think it's fine to say this, especially in a Rust context, and I'd like to merge this PR.

Reply:
I do _not_ strongly disagree would not oppose merging this. (I opted to approve the PR without resolving this conversation thereby trying to express this "I care about this but not enough to block the PR over it." position.)